design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,391 @@
+  containing the connection details should be created into. We could check
+  whether it exists already, but that would be against the behaviour Kubernetes
+  users are used to, e.g. a `Deployment` mounting a `Secret` which does not
+  exist yet is considered valid but remains pending until it’s not created.
+- `publishConnectionDetailsWithStoreConfigRef` : a reference by name to a
+  `StoreConfig`, we could check that it exists, but similarly to the namespace
+  it would go against the behaviour Kubernetes users might be expecting.
+
+### Additional Constraints for the ValidatingWebhook use case
+
+A few constraints that makes validating Compositions from a Validating Webhook
+harder w.r.t. the cases mentioned in the Background:
+
+1. The CRDs of the Managed Resources defined as `base` and traversed by
+   `patches` might not yet have been created when the Composition is being
+   validated.
+2. The CRD of the Composite Resources specified by `CompositeTypeRef` and
+   traversed by `patches` might not yet have been created when the Composition
+   is being validated.
+3. We do not have all the required CRDs locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+In order to address the possible absence of Managed Resources' CRDs, Composite
+Resources' CRDs or any other required external resource at validation time we
+should allow explicitly setting one of 2 modes, `strict` and `loose`(default) .
+These should be configurable through an annotation, e.g.
+`crossplane.io/composition-validation-mode`, on a Composition or maybe even
+globally via a flag, and would respectively imply that any missing resource
+should result in a direct rejection or a possibly incomplete validation.
+
+In `loose` mode we should emit warnings in case any of the preconditions for a
+check are not met, and skip further validation related to that check.
+
+Regarding the approach, I propose to implement a Composition Validating Webhook
+by validating each `resource` and final rendered `Composed
+Resource`, querying the API Server to get the needed schemas.
+
+I suggest we avoid validating `bases`. Doing that would require to ignore
+validating required fields, as those could be set at a later stage by patches
+or transformations. Assuming we implement full `patches` type coverage,
+validating `bases` per se can be avoided, as any error should either be caught
+validating the patches or can be attributed to the `base` itself.
```

Comment:
I agree that full validation of bases can be out of scope (at least at first). Some validation does seem like it would be useful, though. For example:

* Making sure folks aren't setting a field that doesn't exist in the schema (sign of a typo or incorrect indentation).
* Making sure folks aren't setting an unsupported field value (e.g. wrong type, unsupported enum).

Anecdotally I think typos and incorrect indentations in bases are one of the more common things folks run into when writing Compositions.

Reply:
Got it. Per my below comment I didn't realize we had chosen to mock out the XR input. This makes more sense in that light.